File: LinkedList/cp/flatten/flatten.cc
#include "node.h"

class Solution {
 public:
  Node* flatten(Node* head) {
    auto res = dfs(head, true);
    return res.out;
  }

 private:
  typedef struct {
    Node* out;
    Node* np;
  } result;

  result dfs(Node* head, bool topLevel) {
    auto res = result{};
    if (head == NULL) {
      return res;
    }

    auto p = head;
    for (res.np = head; p != NULL; res.np = p, p = p->next) {
      if (topLevel) {
        res.out = head;
      }
      if (p->child != NULL) {
        auto next = p->next;
        auto cHead = p->child;
        p->child = NULL;
        auto child = dfs(cHead, false);
        auto cTail = child.np;

        p->next = cHead;
        cHead->prev = p;
        if (cTail != NULL) {
          cTail->next = next;
          if (next != NULL) {
            next->prev = cTail;
          }
          p = cTail;
        }
      }
    }
    return res;
  }
};

File: linked-list/classic/flatten.go
package classic

func flatten(head *ListNode) *ListNode {
	out, _ := dfs(head, true)
	return out
}

func dfs(head *ListNode, topLevel bool) (out *ListNode, np *ListNode) {
	if head == nil {
		return
	}

	var p *ListNode
	for np, p = head, head; p != nil; np, p = p, p.Next {
		if topLevel {
			out = head
		}

		if p.Child != nil {
			next := p.Next
			cHead := p.Child
			// clear the child
			p.Child = nil
			_, cTail := dfs(cHead, false)
			p.Next = cHead
			cHead.Prev = p
			if cTail != nil {
				cTail.Next = next
				if next != nil {
					next.Prev = cTail
				}
				// jump to the tail of child list
				p = cTail
			}
		}
	}

	return
}
